### PR TITLE
Allow a KMS Action to specify keys in the Resources of a policy

### PR DESCRIPTION
**Pull request by marktheunissen**

### CODE REVIEW

File: policy/resource.go [hunk 1]
```diff
@@ -44,8 +89,18 @@ func (r Resource) isObjectPattern() bool {
...
-		return false
+	if r.isAWSS3() {
+		if strings.HasPrefix(r.Pattern, "/") {
+			return false
+		}
+	}
+	if r.isKMS() {
+		// TODO: Copied from KES repo, confirm that it is generally applicable.
+		if strings.IndexFunc(r.Pattern, func(c rune) bool {
+			return c == '/' || c == '\\' || c == '.'
```
Comment (marktheunissen) [inline policy/resource.go:100]: Is this a good rule for validating key names? Copied from KES.

File: policy/resource.go [hunk 2]
```diff
@@ -26,12 +26,57 @@ import (
...
+const (
+	// ResourceARNPrefix - resource S3 ARN prefix as per AWS S3 specification.
+	ResourceARNPrefix = "arn:aws:s3:::"
+
+	// ResourceARNKMSPrefix is for KMS key resources. MinIO specific API.
+	ResourceARNKMSPrefix = "arn:minio:kms:::"
+)
+
+// ARNPrefixType - ARN prefix type
+type ARNPrefixType uint8
```
Comment (harshavardhana) [inline policy/resource.go:38]:
```suggestion
type ResourceARNType int32

const (
    ResourceARNS3 ResourceARNType = iota
    ResourceARNKMS
    
    unknownARN // for detecting errors
)
```

File: policy/resource.go [hunk 3]
```diff
@@ -44,8 +89,18 @@ func (r Resource) isObjectPattern() bool {
...
 func (r Resource) IsValid() bool {
-	if strings.HasPrefix(r.Pattern, "/") {
-		return false
+	if r.isAWSS3() {
+		if strings.HasPrefix(r.Pattern, "/") {
+			return false
+		}
+	}
+	if r.isKMS() {
+		// TODO: Copied from KES repo, confirm that it is generally applicable.
```
Comment (harshavardhana) [inline policy/resource.go:98]:
```suggestion
```

yes it is acceptable

File: policy/resource.go (hunk 1 above)
Comment (harshavardhana) [reply policy/resource.go:100]: 👍🏼

File: policy/resource.go [hunk 4]
```diff
@@ -136,23 +191,38 @@ func (r Resource) ValidateBucket(bucketName string) error {
...
-	if strings.HasPrefix(pattern, "/") {
+	if strings.HasPrefix(r.Pattern, "/") {
 		return Resource{}, Errorf("invalid resource '%v' - starts with '/' will not match a bucket", s)
 	}
 
+	return r, nil
+}
+
+// NewResourceAWSS3 - creates new resource with the default ARN type of AWS S3.
+func NewResourceAWSS3(pattern string) Resource {
```
Comment (harshavardhana) [inline policy/resource.go:215]:
```suggestion
func NewResourceS3(pattern string) Resource {
```

File: policy/statement.go [hunk 5]
```diff
@@ -129,7 +141,13 @@ func (statement Statement) isValid() error {
...
 
 	if statement.isKMS() {
-		return statement.Actions.ValidateKMS()
+		if err := statement.Actions.ValidateKMS(); err != nil {
+			return err
+		}
+		if err := statement.Resources.ValidateKMS(); err != nil {
+			return err
+		}
+		return nil
```
Comment (harshavardhana) [inline policy/statement.go:150]:
```suggestion
		return statement.Resources.ValidateKMS()
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by marktheunissen** [comment]
Should be good to go, all feedback implemented 👍

File: policy/resourceset.go [hunk 6]
```diff
@@ -150,9 +150,26 @@ func (resourceSet *ResourceSet) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
-// Validate - validates ResourceSet.
-func (resourceSet ResourceSet) Validate() error {
+// ValidateAWSS3 - validates ResourceSet is AWS S3.
+func (resourceSet ResourceSet) ValidateAWSS3() error {
```
Comment (harshavardhana) [inline policy/resourceset.go:154]: don't name any thing AWSS3 its just S3 in all those situations.

File: policy/resourceset.go [hunk 7]
```diff
@@ -150,9 +150,26 @@ func (resourceSet *ResourceSet) UnmarshalJSON(data []byte) error {
...
 	for resource := range resourceSet {
+		if !resource.isAWSS3() {
+			return Errorf("type of resource is not AWS S3 '%v'", resource)
```
Comment (harshavardhana) [inline policy/resourceset.go:157]:
```suggestion
			return Errorf("resource '%v' type is not S3", resource)
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: policy/resource.go [hunk 8]
```diff
@@ -26,12 +26,59 @@ import (
...
+type ResourceARNType uint32
+
+const (
+	// ResourceARNS3 is the ARN prefix type for AWS S3 resources.
+	ResourceARNS3 ResourceARNType = iota + 1
+
+	// ResourceARNKMS is the ARN prefix type for MinIO KMS resources.
+	ResourceARNKMS
+
+	unknownARN // for detecting errors
```
Comment (klauspost) [inline policy/resource.go:47]: Move this to value 0 - then it also catches the zero value. You can use `endARN` here, if you need the size anywhere.

That way `Resource{} ==Resource{Type: unknownARN}`, which seems reasonable.

File: policy/resource.go [hunk 9]
```diff
@@ -136,23 +195,38 @@ func (r Resource) ValidateBucket(bucketName string) error {
...
+	switch {
+	case strings.HasPrefix(s, ResourceARNPrefix):
+		r.Pattern = strings.TrimPrefix(s, ResourceARNPrefix)
+		r.Type = ResourceARNS3
+
+	case strings.HasPrefix(s, ResourceARNKMSPrefix):
+		r.Pattern = strings.TrimPrefix(s, ResourceARNKMSPrefix)
+		r.Type = ResourceARNKMS
+	default:
+		return r, Errorf("invalid resource '%v'", s)
```
Comment (klauspost) [inline policy/resource.go:208]: Simpler and not something we need to maintain:

```suggestion
	for k, v := range ARNPrefixToType {
		if rem, ok := strings.CutPrefix(s, k); ok {
			r.Type = v
			r.Pattern = rem
			break
		}
	}
	if r.Type == unknownARN {
		return r, Errorf("invalid resource '%v'", s)
```

File: policy/resource.go [hunk 10]
```diff
@@ -136,23 +195,38 @@ func (r Resource) ValidateBucket(bucketName string) error {
...
-		return Resource{}, Errorf("invalid resource '%v' - starts with '/' will not match a bucket", s)
+	if strings.HasPrefix(r.Pattern, "/") {
+		return r, Errorf("invalid resource '%v' - starts with '/' will not match a bucket", s)
 	}
 
+	return r, nil
+}
+
+// NewResourceS3 - creates new resource with the default ARN type of AWS S3.
+func NewResourceS3(pattern string) Resource {
```
Comment (klauspost) [inline policy/resource.go:219]: Do we need to break exported API here? Can't we just keep the name. After all the comment says what it does.

File: policy/resource.go [hunk 11]
```diff
@@ -136,23 +195,38 @@ func (r Resource) ValidateBucket(bucketName string) error {
...
 		Pattern: pattern,
-	}, nil
+		Type:    ResourceARNS3,
+	}
 }
 
-// NewResource - creates new resource.
-func NewResource(pattern string) Resource {
+// NewResourceKMS - creates new resource with type KMS
+func NewResourceKMS(pattern string) Resource {
```
Comment (klauspost) [inline policy/resource.go:227]: Very nitpicky: I would say `NewKMSResource` is a bit more natural in terms of language. It does look sillier with all the upper case letters. Feel free to ignore.

**Review by klauspost — COMMENTED** [review]
Just some minor tweaks.

File: policy/resource.go (hunk 8 above)
Comment (marktheunissen) [reply policy/resource.go:47]: Sounds good, but I don't think we need the size? If I add it, I'll need to use it somewhere or disable linting on that line, as the linter complains that it's unused.

File: policy/resource.go (hunk 10 above)
Comment (marktheunissen) [reply policy/resource.go:219]: Sure, I have changed it back.

File: policy/resource.go (hunk 8 above)
Comment (marktheunissen) [reply policy/resource.go]: I've moved to zero value but have not added endARN, LMK if I should and how you would like it used.

**Review by harshavardhana — APPROVED** [review]

**Review by klauspost — APPROVED** [review]

**Pull request merged**